spec.md

```diff
@@ -1090,6 +1180,342 @@ message ControllerServiceCapability {
 
 If the plugin is unable to complete the ControllerGetCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.
 
+#### `CreateSnapshot`
+
+A Controller Plugin MUST implement this RPC call if it has `CREATE_DELETE_SNAPSHOT` controller capability.
+This RPC will be called by the CO to create a new snapshot from a source volume on behalf of a user.
+
+This operation MUST be idempotent.
+If a snapshot corresponding to the specified snapshot `name` already exists and is compatible with the specified `volume_id` in the `CreateSnapshotRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateSnapshotResponse`.
+
+```protobuf
+message CreateSnapshotRequest {
+  // The ID of the source volume to be snapshotted.
+  // This field is REQUIRED.
+  string source_volume_id = 1;
+
+  // The suggested name for the snapshot. This field is REQUIRED.
+  // It serves two purposes:
+  // 1) Idempotency - This name is generated by the CO to achieve
+  //    idempotency. If `CreateSnapshot` fails, the snapshot may or
+  //    may not be created. In this case, the CO may call
+  //    `CreateSnapshot` again, with the same name, to ensure the
+  //    snapshot exists. The Plugin should ensure that multiple
+  //    `CreateSnapshot` calls for the same name do not result in more
+  //    than one snapshot created corresponding to that name. If a
+  //    Plugin is unable to enforce idempotency, the CO's error recovery
+  //    logic could result in multiple (unused) snapshots being created.
+  // 2) Suggested name - Some storage systems allow callers to specify
+  //    an identifier by which to refer to the newly created snapshot.
+  //    If a storage system supports this, it can optionally use this
+  string name = 2;
+
+  // The description of the snapshot.
+  // This field is OPTIONAL.
+  string description = 3;
+
+  // Secrets required by plugin to complete snapshot creation request.
+  // A secret is a string to string map where the key identifies the
+  // name of the secret (e.g. "username" or "password"), and the value
+  // contains the secret data (e.g. "bob" or "abc123").
+  // Each key MUST consist of alphanumeric characters, '-', '_' or '.'.
+  // Each value MUST contain a valid string. An SP MAY choose to accept
+  // binary (non-string) data by using a binary-to-text encoding scheme,
+  // like base64.
+  // An SP SHALL advertise the requirements for required secret keys and
+  // values in documentation.
+  // CO SHALL permit passing through the required secrets.
+  // A CO MAY pass the same secrets to all RPCs, therefore the keys for
+  // all unique secrets that an SP expects must be unique across all CSI
+  // operations.
+  // This information is sensitive and MUST be treated as such (not
+  // logged, etc.) by the CO.
+  // This field is OPTIONAL.
+  map<string, string> controller_create_secrets = 4;
```

Comment:
can we name this to `create_snapshot_secrets`?

too bad that we have `controller_create_secrets` in `CreateVolumeRequest` (should have been `create_volume_secrets`).

Reply:
Yes.

Reply:
Done.